src/hato/client.clj

```diff
@@ -284,42 +275,114 @@
+  [{:keys [retry-exceptions non-retry-exceptions max-retry-count]
+    :or   {retry-exceptions     retry-exceptions
+           non-retry-exceptions non-retry-exceptions
+           max-retry-count      3}}]
+  (fn [_resp ex req retry-count]
+    (and (idempotent-request? req)
+         (some #(instance? % ex) retry-exceptions)
+         (not (some #(instance? % ex) non-retry-exceptions))
+         (< retry-count max-retry-count))))
+
+(def default-retry-handler
+  "A default retry handler which will retry if there is an error with any known retriaretryble
+  exceptions while skipping non-retry exceptions. By default, this will only retry idempotent
+  requests (i.e. GET/HEAD/OPTIONS)."
+  (make-retry-handler {}))
+
+(defn request-with-retries*
+  "Makes a request retrying on failures as defined by retry-handler"
+  [http-client http-request body-handler req retry-handler retry-count]
+  (-> (.sendAsync http-client http-request body-handler)
+      (.handle
+        (reify BiFunction
+          (apply [_ resp e]
+            [resp (unwrap-async-exception e)])))
+      (.thenCompose
+        (reify Function
+          (apply [_ [resp e]]
+            (let [retry-fut (if retry-handler
+                              (retry-handler resp e req retry-count)
+                              (CompletableFuture/completedFuture false))]
+              (-> (if (instance? CompletableFuture retry-fut)
+                    ^CompletableFuture retry-fut
+                    (CompletableFuture/completedFuture retry-fut))
+                  (.thenCompose
+                    (reify Function
+                      (apply [_ retry?]
+                        (if retry?
+                          (request-with-retries* http-client http-request body-handler req retry-handler (inc retry-count))
+                          (if e
+                            (CompletableFuture/failedFuture e)
+                            (CompletableFuture/completedFuture
+                              (response-map {:request     req
+                                             :http-client http-client
+                                             :response    resp}))))))))))))))
+
 (defn request*
-  [{:keys [http-client async? as]
+  [{:keys [http-client async? retry-handler]
     :as   req} & [respond raise]]
   (let [^HttpClient http-client (if (instance? HttpClient http-client) http-client (build-http-client http-client))
-        http-request (ring-request->HttpRequest req)
-        bh (->BodyHandler as)]
-    (if-not async?
-      (let [resp (.send http-client http-request bh)]
-        (response-map
-         {:request     req
-          :http-client http-client
-          :response    resp}))
-
-      (-> (.sendAsync http-client http-request bh)
+        http-request            (ring-request->HttpRequest req)
+        retry-handler           (if (= :auto retry-handler) default-retry-handler retry-handler)
```

Comment:
Should we call it

`:retry-handler` or just `:retry`?

Should we call it `:auto` or `:default`?

Personally, I'd like to make it the retries truly default like Apache HttpClient does, however, I realize that is a decent change to default behavior so opt-in makes sense.